### PR TITLE
HTTP2 Server support

### PR DESCRIPTION
**Pull request by graetzer**
## Scope & Purpose

Support HTTP2 in the server

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)

### Testing & Verification

Until we develop the client part we do not specifcally test the H2 comm task.
Changes to the HttpCommTask are covered by existing testsuites

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8134/
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/9902/

### CODE REVIEW

**Comment by jsteemann** [comment]
Please add an entry for h2 in LICENSES-OTHER-COMPONENTS.md

File: arangod/GeneralServer/H2CommTask.cpp [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+    // we need a continuous block of memory for headers
+    std::vector<nghttp2_nv> nva;
+    nva.reserve(4 + res.headers().size());
+
+    std::string status = std::to_string(static_cast<int>(res.responseCode()));
+    nva.push_back({(uint8_t*)":status", (uint8_t*)status.data(), 7,
+                   status.size(), NGHTTP2_NV_FLAG_NO_COPY_NAME});
+
+    bool seenServerHeader = false;
+    // bool seenConnectionHeader = false;
```
Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.cpp]: not needed

File: arangod/GeneralServer/H2CommTask.cpp [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+        size_t nread = std::min(length, body.size() - strm->responseOffset);
+        if (nread == 0 || strm->responseOffset == body.size()) {
+          *data_flags |= NGHTTP2_DATA_FLAG_EOF;
+        } else {
+          const char* src = body.data() + strm->responseOffset;
+          std::copy_n(src, nread, buf);
+          strm->responseOffset += nread;
+        }
+        
+        // simon: might be needed if NGHTTP2_DATA_FLAG_NO_COPY is used
```
Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.cpp]: remove?

File: arangod/GeneralServer/H2CommTask.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  static int on_frame_send(nghttp2_session* session, const nghttp2_frame* frame,
+                           void* user_data);
+
+  static int on_frame_not_send(nghttp2_session* session, const nghttp2_frame* frame,
+                               int lib_error_code, void* user_data);
+
+ private:
+  // ongoing Http2 stream
+  struct Stream {
+    Stream(std::unique_ptr<HttpRequest> req) : request(std::move(req)) {}
```
Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.h]: let's make this explicit, to pacify cppcheck

File: arangod/GeneralServer/H2CommTask.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  Stream* createStream(int32_t sid, std::unique_ptr<HttpRequest>);
+  Stream* findStream(int32_t sid) const;
+
+  // may be used to signal a write from sendResponse
+  void signalWrite();
+
+ private:
+  static constexpr size_t kOutBufferLen = 64 * 1024 * 1024;
+  std::array<uint8_t, kOutBufferLen> _outbuffer;
```
Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.h]: Does this mean we have a fixed size buffer of 64MB per H2CommTask? That seems a bit much. Can we allocate this buffer on the heap if needed?

File: arangod/GeneralServer/H2CommTask.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  // no more than 64 streams allowed
+  boost::lockfree::queue<HttpResponse*, boost::lockfree::capacity<64>> _responses;
+
+  std::map<int32_t, std::unique_ptr<Stream>> _streams;
```
Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.h]: does this map need to be sorted? how many streams do we expect on average?

File: arangosh/Shell/V8ClientConnection.cpp [hunk 6]
```diff
@@ -94,6 +94,8 @@ V8ClientConnection::V8ClientConnection(application_features::ApplicationServer&
 V8ClientConnection::~V8ClientConnection() {
   _builder.onFailure(nullptr);  // reset callback
   shutdownConnection();
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp]: is 100 milliseconds a safe bet here? :cry:

File: arangod/GeneralServer/H2CommTask.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/GeneralServer/H2CommTask.cpp]: I left it in for future reference

File: arangod/GeneralServer/H2CommTask.h (hunk 5 above)
Comment (graetzer) [reply arangod/GeneralServer/H2CommTask.h]: https://chromium.googlesource.com/chromium/src/+/master/base/containers/README.md#Map-and-set-selection-Usage-advice
> Generally avoid std::unordered_set and std::unordered_map. In the common case, query performance is unlikely to be sufficiently higher than std::map to make a difference, insert performance is slightly worse, and the memory overhead is high. This makes sense mostly for large tables where you expect a lot of lookups.

Comment (graetzer) [reply arangod/GeneralServer/H2CommTask.h]: the max is 64

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/8173/

File: arangod/GeneralServer/H2CommTask.cpp [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+       //      << GeneralRequest::translateMethod(::llhttpToRequestType(&_parser))
+       << "\",\"" << static_cast<int>(res->responseCode()) << "\","
+       << Logger::FIXED(totalTime, 6);
+  
+  // this uses fixed capacity queue, push might fail
+  int retries = 1024;
+  while (!_responses.push(static_cast<HttpResponse*>(res.get()))) {
+    std::this_thread::yield();
+    if (--retries == 0) {
+      return;
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/GeneralServer/H2CommTask.cpp]: added a stream RST now

File: arangod/GeneralServer/H2CommTask.cpp [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+
+  // https://http2.github.io/http2-spec/#discover-http
+  auto str = std::make_shared<std::string>();
+  str->append(
+      "HTTP/1.1 101 Switching Protocols\r\nConnection: "
+      "Upgrade\r\nUpgrade: h2c\r\n\r\n");
+
+  asio_ns::async_write(this->_protocol->socket, asio_ns::buffer(str->data(), str->size()),
+                       [self(this->shared_from_this()),
+                        str(std::move(str)),
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/GeneralServer/H2CommTask.cpp]: fixed

File: 3rdParty/fuerte/src/loop.cpp [hunk 9]
```diff
@@ -62,5 +54,14 @@ asio_ns::ssl::context& EventLoopService::sslContext() {
   }
   return *_sslContext;
 }
+
+void EventLoopService::stop() {
+  // allow run() to exit, wait for threads to finish only then stop the context
+  std::for_each(_guards.begin(), _guards.end(), [](auto& g) { g.reset(); });
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
```
[the remark this answers is not in the document]
Comment (graetzer) [reply 3rdParty/fuerte/src/loop.cpp]: seems to work fine now, perhaps we can even remove them in the next PR

**Pull request merged**